### DISCUSSION TITLE
Mac/Linux Users - come in to discuss on how to use these modules

### DISCUSSION THREAD

**Discussion by ozbillwang** [Azure/ResourceModules#1237] [Show and tell]
As Mac/Linux user, this repo is hard for us to use directly, we need do some converts, and more examples. 

similar idea has been raised in this PR (https://github.com/Azure/ResourceModules/pull/1236) 

Let's move forward, I share a sample to create virtual network for Mac/Linux users. 

### Sample with simple bicep module

```
env="DEV"
namePrefix="ExampleDeployment"
name="${namePrefix}-${env}"
resourceGroup="ExampleGroup"
region="Central US"

az group create --name "${resourceGroup}" --location "${region}"
inputObject=(
    '--name'           "${name}-VNET"
    '--resource-group' "${resourceGroup}"
    '--template-file'  "${home}/ResourceModules/arm/Microsoft.Network/virtualNetworks/deploy.bicep"
    '--parameters'     'addressPrefixes=["10.1.0.0/16"]'
    '--parameters'     name="${name}-VNET"
)
az deployment group create ${inputObject[@]}
```

### Sample with multiple bicep modules, Using one module's output to be referenced by other modules.

I made a repo to show a real example, on how I use this azure official bicep modules to create a virtual network and virtual machine scale sets in it

https://github.com/ozbillwang/Azure-Bicep-Best-Practices

It should save you a lot of time to jump in with these bicep modules easily, if you can follow up my codes.

**Comment by ozbillwang** [comment]
can't work it out with parameters key/value directly for `subnets`, and only work it out with parameter files.

These are not working:

```
'--parameters'     subnets=[{"name":"AppSubnet"], "addressPrefix":"10.1.0.0/24"}]
```

or
```
'--parameters'     subnets=[{\"name\":\"AppSubnet\"], \"addressPrefix\":\"10.1.0.0/24\"}]
```

**Comment by ozbillwang** [comment]
I'd like to use the utility [SetModuleReadMe](https://github.com/Azure/ResourceModules/wiki/UtilitiesSetModuleReadMe) to generate README for my own bicep. But got error. 

Here I show all steps for reference

* cd to the folder where your bicep file is located
* git clone git@github.com:Azure/ResourceModules.git
* run `PowerShell` in docker image, because we need run it on Mac/Linux
* install az cli 
* start powershell (`pwsh`)
* load module `Set-ModuleReadMe.ps1`
* generate the readme

```
docker run -ti --rm -v $(pwd):/apps -w /apps --entrypoint=bash mcr.microsoft.com/powershell
root@0b1271c2e575:/apps# apt update
root@0b1271c2e575:/apps# apt install curl
root@0b1271c2e575:/apps# curl -sL https://aka.ms/InstallAzureCLIDeb | bash
root@0b1271c2e575:/apps# pwsh
PS /apps> . /apps/ResourceModules/utilities/tools/Set-ModuleReadMe.ps1
PS /apps> Set-ModuleReadMe ./deploy.bicep
```

But got error
```
InvalidOperation: /apps/utilities/tools/Set-ModuleReadMe.ps1:556
Line |
 556 |          $serviceIdentifiers = $fullResourcePath.Replace('Microsoft.', …
     |          ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
     | You cannot call a method on a null-valued expression.

InvalidOperation: /apps/utilities/tools/Set-ModuleReadMe.ps1:557
Line |
 557 |  … s | ForEach-Object { $_.substring(0, 1).toupper() + $_.substring(1) }
     |                         ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
     | You cannot call a method on a null-valued expression.

InvalidOperation: /apps/utilities/tools/Set-ModuleReadMe.ps1:161
Line |
 161 |  … ategories = $descriptions | ForEach-Object { $_.Split('.')[0] } | Sel …
     |                                                 ~~~~~~~~~~~~~~~~
     | You cannot call a method on a null-valued expression.

Merge-FileWithNewContent: /apps/utilities/tools/Set-ModuleReadMe.ps1:227
Line |
 227 |  … -oldContent $ReadMeFileContent -newContent $newSectionContent -Sectio …
     |                                               ~~~~~~~~~~~~~~~~~~
     | Cannot bind argument to parameter 'newContent' because it is an empty array.

Set-OutputsSection: /apps/utilities/tools/Set-ModuleReadMe.ps1:627
Line |
 627 |          $readMeFileContent = Set-OutputsSection @inputObject
     |                                                  ~~~~~~~~~~~~
     | Cannot bind argument to parameter 'ReadMeFileContent' because it is null.

Set-TemplateReferencesSection: /apps/utilities/tools/Set-ModuleReadMe.ps1:637
Line |
 637 |  …       $readMeFileContent = Set-TemplateReferencesSection @inputObject
     |                                                             ~~~~~~~~~~~~
     | Cannot bind argument to parameter 'ReadMeFileContent' because it is null.

Set-TableOfContent: /apps/utilities/tools/Set-ModuleReadMe.ps1:646
Line |
 646 |          $readMeFileContent = Set-TableOfContent @inputObject
     |                                                  ~~~~~~~~~~~~
     | Cannot bind argument to parameter 'ReadMeFileContent' because it is null.

VERBOSE: File [./readme.md] updated
```

**Reply by AlexanderSehr** [reply]
Hey @ozbillwang, regarding the examples for different OS types I agree - we can extend them. 

Regaring the error messages below I can at least say that all 'ReadMeFileContent' related errors stem from the fact that the function was not able to properly read the template.

That this line `$serviceIdentifiers = $fullResourcePath.Replace('Microsoft.` fails is curious though. Right before it replaces all slashes to a common approach so it 'should' work. Not sure how I can reproduce the error right now.

**Comment by segraef** [comment]
Happy to support and test snippets on MacOS under `pwsh` if needed.

**Reply by segraef** [reply]
[Register](https://github.com/Azure/ResourceModules/blob/main/utilities/tools/AzureDevOps/Register-AzureDevOpsPipeline.ps1) seems to be tricky as well under MacOS with pwsh.

<img width="840" alt="image" src="https://user-images.githubusercontent.com/19261257/181228706-22b173ce-b15f-466e-b968-f58daaea92b0.png">
